lib/column/variant.go

```diff
@@ -0,0 +1,385 @@
+		}
+
+		return nil, nil
+	default:
+		if valuer, ok := v.(driver.Valuer); ok {
+			val, err := valuer.Value()
+			if err != nil {
+				return nil, &ColumnConverterError{
+					Op:   "Append",
+					To:   string(c.chType),
+					From: fmt.Sprintf("%T", v),
+					Hint: "could not get driver.Valuer value",
+				}
+			}
+
+			return c.Append(val)
+		}
+
+		return nil, &ColumnConverterError{
+			Op:   "Append",
+			To:   string(c.chType),
+			From: fmt.Sprintf("%T", v),
+		}
+	}
+}
+
+func (c *Variant) AppendRow(v any) error {
+	var requestedType string
+	switch v.(type) {
+	case nil:
+		c.appendNullRow()
+		return nil
+	case chcol.Variant:
+		v = v.(chcol.Variant).Any()
+		if v == nil {
+			c.appendNullRow()
+			return nil
+		}
+	case *chcol.Variant:
+		v = v.(*chcol.Variant).Any()
+		if v == nil {
+			c.appendNullRow()
+			return nil
+		}
+	case chcol.VariantWithType:
+		requestedType = v.(chcol.VariantWithType).Type()
+		v = v.(chcol.VariantWithType).Any()
+		if v == nil {
+			c.appendNullRow()
+			return nil
+		}
+	case *chcol.VariantWithType:
+		requestedType = v.(*chcol.VariantWithType).Type()
+		v = v.(*chcol.VariantWithType).Any()
+		if v == nil {
+			c.appendNullRow()
+			return nil
+		}
+	}
+
+	if requestedType != "" {
+		typeIndex, ok := c.columnTypeIndex[requestedType]
+		if !ok {
+			return fmt.Errorf("value %v cannot be stored in variant column %s with requested type %s: type not present in variant", v, c.chType, requestedType)
+		}
+
+		if err := c.columns[typeIndex].AppendRow(v); err != nil {
+			return fmt.Errorf("failed to append row to variant column with requested type %s: %w", requestedType, err)
+		}
+
+		c.appendDiscriminatorRow(typeIndex)
+		return nil
+	}
+
+	// If preferred type wasn't provided, try each column
+	var err error
+	for i, col := range c.columns {
+		if err = col.AppendRow(v); err == nil {
+			c.appendDiscriminatorRow(uint8(i))
+			return nil
+		}
+	}
+
+	return fmt.Errorf("value \"%v\" cannot be stored in variant column: no compatible types", v)
+}
+
+func (c *Variant) encodeHeader(buffer *proto.Buffer) {
+	buffer.PutUInt64(SupportedVariantSerializationVersion)
+}
+
+func (c *Variant) encodeData(buffer *proto.Buffer) {
+	buffer.PutRaw(c.discriminators)
+
+	for _, col := range c.columns {
+		col.Encode(buffer)
+	}
+}
+
+func (c *Variant) Encode(buffer *proto.Buffer) {
+	c.encodeHeader(buffer)
+	c.encodeData(buffer)
+}
+
+func (c *Variant) ScanType() reflect.Type {
+	return scanTypeVariant
+}
+
+func (c *Variant) Reset() {
+	c.discriminators = c.discriminators[:0]
+
+	for _, col := range c.columns {
+		col.Reset()
+	}
+}
+
+func (c *Variant) decodeHeader(reader *proto.Reader) error {
+	variantSerializationVersion, err := reader.UInt64()
+	if err != nil {
+		return fmt.Errorf("failed to read variant discriminator version: %w", err)
+	} else if variantSerializationVersion != SupportedVariantSerializationVersion {
+		return fmt.Errorf("unsupported variant discriminator version: %d", variantSerializationVersion)
+	}
```

Comment:
nit 
```suggestion
	if err != nil {
		return fmt.Errorf("failed to read variant discriminator version: %w", err)
	} 
	
	if variantSerializationVersion != SupportedVariantSerializationVersion {
		return fmt.Errorf("unsupported variant discriminator version: %d", variantSerializationVersion)
	}
```

IMO more readable